syntax = "proto3";

package chatbot;

// Chatbot service with streaming support
service ChatbotService {
  // Stream chat messages for real-time conversation
  rpc StreamChat(stream ChatMessage) returns (stream ChatResponse);
  
  // Single request/response for simple queries
  rpc ProcessQuery(ChatMessage) returns (ChatResponse);
  
  // Get available tools
  rpc GetAvailableTools(Empty) returns (ToolsResponse);
}

// Chat message from client
message ChatMessage {
  string query = 1;
  string session_id = 2;
  string user_id = 3;
  map<string, string> metadata = 4;
}

// Chat response from server
message ChatResponse {
  string response_id = 1;
  string session_id = 2;
  ResponseType type = 3;
  string content = 4;
  ToolExecution tool_execution = 5;
  SearchResult search_result = 6;
  ErrorInfo error_info = 7;
  bool is_final = 8;
  float confidence = 9;
  string reasoning = 10;
}

// Tool execution information
message ToolExecution {
  string tool_name = 1;
  string service = 2;
  string search_type = 3; // "knowledge_graph" or "vector_search"
  bool success = 4;
  map<string, string> parameters = 5;
}

// Search result details
message SearchResult {
  string search_method = 1;
  int32 entities_found = 2;
  int32 relationships_found = 3;
  int32 chunks_retrieved = 4;
  repeated string entities = 5;
  repeated string relationships = 6;
}

// Error information
message ErrorInfo {
  string error_code = 1;
  string error_message = 2;
  string error_details = 3;
}

// Available tools response
message ToolsResponse {
  repeated Tool tools = 1;
  int32 total_tools = 2;
}

// Tool information
message Tool {
  string tool_name = 1;
  string description = 2;
  string category = 3;
  repeated string examples = 4;
  map<string, ParameterInfo> parameters = 5;
}

// Parameter information
message ParameterInfo {
  string type = 1;
  string description = 2;
  bool required = 3;
}

// Empty message for simple requests
message Empty {}

// Response types
enum ResponseType {
  GENERAL_RESPONSE = 0;
  TOOL_EXECUTION = 1;
  SEARCH_RESULT = 2;
  ERROR = 3;
  MANAGEMENT_SUPPORT = 4;
  NO_TOOL_FOUND = 5;
  STREAMING_CHUNK = 6;
  STREAMING_COMPLETE = 7;
}

